### ISSUE TITLE
Nodes must be linserted in the ascending time stamp order

### ISSUE THREAD

**Issue by IzzySoft** [StoneStepsInc/StoneStepsWebalizer#10]
I just noticed on one of my logs I get the above error; SSW refuses to process that file. Looking inside (after having fetched the "Log time" from the db) I see that Apache indeed inserts a line out of sequence every now and then, especially if on "high load" it seems. Is there a way to

* tell SSW to simply skip the out-of-sequence lines?
* tell SSW to skip to a given later time stamp (provided it is a single hick-up)?
* some other way to deal with that, not losing half a week's stats?

cutting/editing the log is not really an option with a file size of > 240M. I neither have a way telling Apache to "not do that".

**Comment by IzzySoft** [comment]
PS: I've just sorted the entire log by its "time column", so there should be no more lines "out of sequence". The error message persists. Strange. Unfortunately, not even passing the `-d` parameter tells me which log line it is complaining about (actually, `-d` does not reveal *any* additional details, while I had expected a lot of debug info – neither did `Debug yes` in the config, which I just tried).

**Comment by StoneStepsInc** [comment]
Took me some time to go through history here. This was the change to manage the amount of memory used when processing large logs, so it would save data from older log records to disk and use memory only for new data. As a troubleshooting step, you can try increasing `DbCacheSize` to some high value, which might be able to keep it all in memory, but it's not guaranteed to help, since we don't know what's going on under the covers.

This error is a safeguard and isn't supposed to be reported for out-of-order log records, which should be just reported and ignored, so it looks like something is broken there. If you can give me this log file for testing, email me the link where I can pick it up and I will have a look at it.

**Comment by IzzySoft** [comment]
Thanks Andrew! I've tried setting DbCacheSize to 2G (`DbCacheSize 2000M`) to no avail, error remains. I'll send you a link by mail.

**Comment by StoneStepsInc** [comment]
I'm afraid I cannot reproduce this. I will give it more thought, but wanted to post an update for the time being.

I see that you used v6.0.0, so I tested it with the same version on a Debian/x64 VM. Here's the `--db-info` output for your database:
```
Created by      : 6.0.0.7
Last updated by : 6.0.0.7
First day       : 2021/05/01
Log time        : 2021/05/11 15:17:04
Active visits   : 688
Active downloads: 0
Incremental     : yes
Batch           : no
Local time      : yes
UTC offset      : 0 min
Numeric storage : c=1 s=2 i=4 l=8 ll=8 d=8
Byte order      : 78563412
Byte order x64  : EFCDAB9078563412
```
The way it works is that when it reads the database, it uses the current log time, which is `2021/05/11 15:17:04` in this case, to load all active visits and their hosts and URLs from `webalizer.db`. That's the base time stamp before it starts.

After this it starts processing log files and it will skip all log records with a time stamp less than or equal to the database time stamp.

Once the first log record with a time stamp greater then the database time is found, the equal check is disabled and it will ignore only log records with a time stamp that is less than the current one. For your log file it's still quite a few log records - about 69K out of 2M, but they are just recorded as ignored. This can only be remedied by sorting log files before processing, I'm afraid.

Processing your log with the database, it correctly skips all log records with time stamps less than or equal to `2021/05/11 15:17:04` and finishes the rest, ignoring shuffled records as it goes along.

So, shuffled time stamps are being caught by the log processor and don't reach the error you are seeing and I cannot quite figure out a possible condition how it would get around the log record time check before it starts updating the database.

I wonder if somehow a time zone gets in the way here, but I don't see any direct connection so far. Your time stamps are local and it handles them as such without any conversions. Mentioning this just as an angle, not saying it's in this area.

So, somehow in your setup it manages to get around the log time stamp check and attempts to insert one of the time-tracked records (e.g. host) with a lower log record time stamp than already exists in the database. One bigger area I didn't go through with a fine-tooth comb is GeoIP - wonder if any database updates happen with different timing in your setup. Will do this later today and will post a bit more.

**Comment by StoneStepsInc** [comment]
I cannot see it caused by GeoIP code either, but then again, I'm just looking at the code without debugging. If you can build from the source, you can use this patch to build a test executable that prints time stamps and the source type of the offending record.
```
diff --git a/src/hashtab_tmpl.cpp b/src/hashtab_tmpl.cpp
index 746e3e4..5df2a6e 100644
--- a/src/hashtab_tmpl.cpp
+++ b/src/hashtab_tmpl.cpp
@@ -147,8 +147,11 @@ node_t *hash_table<node_t>::put_node(uint64_t hashval, node_t *node, int64_t tst
    }
    else {
       // enforce time stamp order for new regular nodes
-      if(!tmlist.empty() && tmlist.back()->tstamp > tstamp)
-         throw std::logic_error("Nodes must be linserted in the ascending time stamp order");
+      if(!tmlist.empty() && tmlist.back()->tstamp > tstamp) {
+         throw std::logic_error(string_t::_format(
+                     "Nodes must be inserted in the ascending time stamp order (%" PRIx64 ", %" PRIx64 ", %s)",
+                     tmlist.back()->tstamp, tstamp, typeid(node).name()));
+      }
 
       //
       // Insert a new list node with a nullptr value at the end of the list and allocate 
```
If your environment isn't setup for this (can be a hassle with dependencies), you can probably run the binary I attached to this post. It is built on my fairly old Debian 9 x64 VM. If you need some other flavor, let me know and I'll see if I can get a suitable VM.

[webalizer-debian-9-x64-node-ts-err.gz](https://github.com/StoneStepsInc/StoneStepsWebalizer/files/6489763/webalizer-debian-9-x64-node-ts-err.gz)

This test output is quite limited, but it would help to know which time stamps it fails on and for which item type.

**Comment by IzzySoft** [comment]
Thanks Andre! Meanwhile logrotate has kicked in and, as I expected, the new log was picked up fine by SSW – leaving a 4d gap in the stats. So I copied the 2 saved files (`.db` and `.hist`) to a new temporary output dir, created an adjusted config using that, and ran the binary you've attached against that. Output looks much identical to what I had before – with the single difference that it now mentions the nodes:

    Nodes must be inserted in the ascending time stamp order (609a84e8, 0, P10storable_tI7anode_tE)

Can you identify a log line by that? My somehow wild guess is that if I'd remove all but one of those `2021/05/11 15:17:04` lines (and maybe the next 10 along), it would work. Eh, nice guess – that would have been too easy, no dice. Not even removing the next 100 lines did help. Deleting all lines equal or older the given timestamp (faking a logrotate) just makes the error showing promptly.

So as it started working fine again after the log switch, it must be some entry after the given timestamp. But I've got no idea which or how to find out. It must be a record within the next hour of that (as processing runs hourly). Interestingly, removing the next hour gives the same abort, just with a different address:

    Nodes must be inserted in the ascending time stamp order (609bd8b4, 0, P10storable_tI7anode_tE)

To that address it then sticks for hours (I went on chopping the log by 1h each time – big thanks to `vi`: just go to the first line you want to keep, then issue `dgg` and all above is gone). Then, having reached `12/May/2021:18:18:01` (meanwhile chopping 2h at once), address changes once more:

    Nodes must be inserted in the ascending time stamp order (609d63d9, 0, P10storable_tI7anode_tE)

New address at `13/May/2021:20:18:00`:

    Nodes must be inserted in the ascending time stamp order (609eaa0e, 0, P10storable_tI7anode_tE)

This address then sticks until I reach `14/May/2021:20:18:00` – from there, SSW picks up fine again.

I've no idea what those addresses tell – but I hope they help you finding out what the problematic places might have in common. I thought about rerunning my stats from there (simply replacing the two files as they are now and have SSW pick up from there with the new log should work, right?) – but I'm not sure if 1 day of stats is worth it should it fail :see_no_evil:

Would it help you to have my configs? Then I'd bundle them up for you as well. Who knows, maybe some specific settings do not work that well together…

**PS:** I still kept the "original copy" (log + those 2 files) so I can test again if you think it might help. I don't care too much about that gap, but gladly help fixing that problem.

**Comment by StoneStepsInc** [comment]
Sorry, I have to run now and will read more into this, but this already provides some additional data that the problem is with agent records (`anode_t`) and the time stamp in the look-up structure is for 2021-11-05 13:21:44 UTC (that's the last agent it inserted) and the one it's trying to insert at the moment of failure doesn't have a time stamp (the zero is for 1970-01-01 00:00:00 UTC), which certainly gives me more food for thought.

Will write more later. Thanks for the additional details.

**Comment by IzzySoft** [comment]
> the one it's trying to insert at the moment of failure doesn't have a time stamp

That's funny. I've just sorted the log by timestamp (`sort -k 3 access_log > access_log_sorted`). If there were an entry with no date, or with `1970-01-01 00:00:00 UTC`, it must be the very first row. No such thing – but you can check that for yourself, having the log. I see the wheels spinning in your head, indeed :see_no_evil:

Whenever time permits, of course. I'll stand by for testing. Thanks a lot!

**Comment by StoneStepsInc** [comment]
Before I forget, don't forget to replace the test binary with the one from the package. It may contain small changes from the main branch. Shouldn't change anything, but it's safer to go with the one that went through some testing.

After looking at a bunch of code for q bit, I can see the only logical explanation that can cause this behavior. If there is a `GroupAgent` directive that has the exact same value as an actual user agent name from the log and if a group with the same name was added after some data for the month has been processed, then it is possible to see this error.

It basically saves some user agent into the database as a regular (non-group) item on the first run and on a subsequent run finds the new `GroupAgent` entry and attempts to cache the group name in memory with a zero time stamp (i.e. those in debug output lines you posted), but in the process finds that regular user agent name saved earlier and tries to restore the user agent as a regular entry, not a group, but the time stamp is missing (i.e. just a zero), so it reports this error.

I don't know which user agent does it - I got all this just analyzing the code and not debugging, so no specifics. All I can say is that `2021-11-05 13:21:44 UTC` is the last valid user agent entry (the first time stamp in the debug output as a Unix epoch time stamp, in hex), so one of the following log lines is probably the culprit, but it's hard to say how far it would be from that point - depends on how many unique user agents follow this line.

It should work fine if you don't change configuration in the middle of the month, which would distort stats anyway (e.g. new group wouldn't get all of its hits), so if you change configuration, it's better to reprocess all logs for this month to get consistent counts.

Let me know if you find which group is causing this. I will keep the bug open for now.

**Comment by IzzySoft** [comment]
> Before I forget, don't forget to replace the test binary with the one from the package.

No worries. I've run that one from a temporary location, not touching the one installed "properly".

> If there is a GroupAgent directive that has the exact same value as an actual user agent name from the log

That could well be – I see e.g. a plain "Mozilla" in the log.

> and if a group with the same name was added after some data for the month has been processed

Nope. That part of the config was last touched `Jun 28  2020`. Thanks to my modular config, that's easy to tell.

> I don't know which user agent does it

Couldn't the debug version spit that out when it happens? That wouldn't spam too much as it's an exception: it dumps the record once, and then exits anyway.

> It should work fine if you don't change configuration in the middle of the month

Well, I did not now – last change to config was in April.

> Let me know if you find which group is causing this.

With that extra debug, I might be able to. As I didn't change any group lately, situation must be a little different. Again, maybe it helps if I send you my config?

**Comment by StoneStepsInc** [comment]
> Nope. That part of the config was last touched Jun 28 2020. Thanks to my modular config, that's easy to tell.

That's the only condition under which this message can be reported - if the user agent in the database is a non-group agent and the one from the log is considered a group agent.

> Couldn't the debug version spit that out when it happens? 

That code handles all record types, so it's not as easy to do it there, which is why debug details are fairly terse. Now that I know it's an agent, I can add some debug code for that and it can tell which instead of the message you are seeing, but I will have to find some time in the week for that, so will be a few days.

I'll post here when I have that debug build with detection of a group/non-group flip.

**Comment by IzzySoft** [comment]
Thanks – I'll gladly test the binary again then. Btw: you should have received my config as a tarball, with all timestamps etc. preserved, so you can see yourself which part was last changed when. The corresponding "site" config just has a recent date because of the current testing…

**Comment by StoneStepsInc** [comment]
Yes, I saw an email, but haven't had a chance to read it. Thank you for the extra details. Will post more later. Need to juggle priorities during the week.

**Comment by StoneStepsInc** [comment]
Here's a binary with a quick change that will report the name of the user agent if it finds one in the state database that is different from the one found in the log. It will no longer report those other debug messages, but rather something similar to this:
```
Inconsistent grouping used for a user agent (Brainstorm)
```
, except that (hopefully) the name will be the one that is causing these troubles.

If this doesn't provide enough detail, it will have to wait until the weekend, I'm afraid. Let me know. Hopefully it will work out.

[webalizer-ua-grp-chk-6-0.gz](https://github.com/StoneStepsInc/StoneStepsWebalizer/files/6497437/webalizer-ua-grp-chk-6-0.gz)

I will see if I can put this check in all code that can end up reading from the database something that was grouped differently before.

**Comment by IzzySoft** [comment]
Inconsistent grouping used for a user agent (F-Droid)

Indeed there is a group by this name – and I've seen log entries going by this name (though this agent usually carries a version number (e.g. `"F-Droid 1.10"`). Running the binary on the combi (log, db, conf) I've sent to you:

    Inconsistent grouping used for a user agent (Mozilla)

Again: yes, that group exists (for years actually, as it's the fallback for all those using "Mozilla" but have no better match). It's even part of the original config (dated 6/2008) – so definitely nothing I changed in the config during the month. My guess is each of the addresses/hashes from above

    Nodes must be inserted in the ascending time stamp order (609a84e8, 0, P10storable_tI7anode_tE)

corresponds to a different UA, so we have 2 or 3 more ("F-Droid" is from the new log file).

Apart from getting this more precise message into the logs, the question is how to remedy the issue. Could an override be used – like when this situation arises, the group has preference and the separate UAs get "merged" into the group (if needed, including the historical ones) if such a conflict shows up?

**Comment by StoneStepsInc** [comment]
> My guess is each of the addresses/hashes from above

Just to be clear, those are time stamps, in hex, of the most recent unique item, such as a user agent, added to a look-up structure, so `609a84e8` > `1620739304` > `2021-05-11 13:21:44 UTC`.

Group names and item names share the same key space in look-up structures and are just marked as regular items or group names in the data. This means there's no way to find separately an individual item or a group name. It's always been this way and the assumption always was that they are different.

What you are highlighting, however, is something I need to look into, which is that without changing configuration, somehow an item is identified as a regular item first and saved after the first run as such and then in a subsequent run is identified as a group and this switch of type occurs that results in the error you observed.

I will run a few tests to see how this can happen without changing configuration and will give grouping some thought in general. Will post here once I have something.

**Comment by IzzySoft** [comment]
Should you need the logs back to the start of the month, just let me know and I ready them for your pickup. That would be ~80M compressed.

**Comment by StoneStepsInc** [comment]
I can confirm that it's not a change in configuration that causes it, but rather the fact that it adds items first under whatever name they have and then does grouping, so when there's a group name matches exactly the item name, the item name gets stored in the database and on the second run triggers this error.

Problem is, however, that even when it doesn't report the error, it miscounts matching group and item names. That is, it will find it once as an item and will bump up the count and then it will find it as a group and will do the same, but for the item, so the report will be misleading. What's displayed depends on what was seen first - item or the group. Most of the times the former, but I can see configurations where the latter may also happen.

This behavior goes back to the time I forked the project, except that it may not have reported it and just counted as described, although I didn't run prior versions - just looked at the code, so may be something is different a bit from what's described.

I will think how to deal with this, as I certainly don't want to go back to miscounting items vs. groups. Will post more later.

**Comment by IzzySoft** [comment]
Good to read the culprit has been identified! Just wondering why it didn't pop up earlier here (and while it does now, just for a single site (luckily)). Never mind, more important is to look forwards. So I'm looking forward to your next news :smiley: Thanks!

**Comment by StoneStepsInc** [comment]
Just to answer your question, it wouldn't generate an error, bit rather would yield incorrect counts, so unless you noticed those discrepancies for affected user agents and other items, it would just run unnoticed. It should be easy to reproduce, once you know what to look for - have two log files with one line in each with a user agent that matches some group name and process these logs in two separate runs - you should see that the user agent will have 2 hits and group probably won't be there.

The above is the expectation from just looking at the code. Once I figure out how to deal with it, I will capture more carefully what actually is going on besides of the generated error you are seeing.

**Comment by IzzySoft** [comment]
Got it, thanks! My problem at the moment is that it generates no stats at all until the next logrotate, then generates stats for maybe half a day (like this time) and after that has another 6.5d gap :see_no_evil: But no worries, unless it takes you more than a month or two, I should be able to reset (to the `.db` and `.hist` I've saved) and re-run all logs since the start of the first gap. Nothing life-threatening. Glad to be able to provide you some "testing material" :rofl:

**Comment by StoneStepsInc** [comment]
I investigated this a bit more and wanted to clarify the incorrect statement I made earlier in that matching group names and regular items would cause miscounts going back a few versions. It only happened in the database mode, which wasn't the default processing mode. The default mode was to handle everything in memory, so this bug would never happen. At some point in v5 I improved the database mode to handle large logs with reasonable amount of memory and didn't consider the case where group names matched item names.

So, the bug is in failing to compare the group/non-group indicator when reading existing data from the database. I will work on it this weekend and will post at some point where I am with this. It will most likely take me a few days to fix it. It's not quite trivial, unfortunately.

**Comment by IzzySoft** [comment]
Thanks for the heads-up Andre! I'll be AFK for most of the time over the next few days anyway (busy with some RL stuff), so I'm not able to test things before Monday. Best luck for fixing!

**Comment by StoneStepsInc** [comment]
I implemented a simpler fix that has a negligible performance cost for items with same values and group names, like `Mozilla` in your logs. It won't be noticeable when processing typical logs (i.e. if you have a large log only with same-value user agents that match group names, then perhaps), but should be backward compatible with the existing databases, so you can just pick up from the last good database and process outstanding logs.

You can grab the package in Releases:

https://github.com/StoneStepsInc/StoneStepsWebalizer/releases/tag/v6-0-1

I tested this fix in my setup and it appears to be working, but I will keep this issue open until you had a chance to confirm that it's working on your end as well. Let me know.

**Comment by IzzySoft** [comment]
Thanks a lot, works fine! :tada: Just replaced the original `webalizer.db` plus `webalizer.hist` and re-processed the files having left gaps, all ran threw fine and the report looks as expected.

Two questions:

* releases now have a separate Ubuntu package. What's the difference to the Debian one I packaged until now? Asking because my repo is otherwise mostly distro-agnostic.
* looks like the `sample.conf` is still left without the additional explanations. I don't remember if we already merged mine – ah, yes, we did with #9 – so I wonder why those didn't make it into the tarball(s) and whether I should package your original file or my copy with the additions?

**Comment by StoneStepsInc** [comment]
Thanks for confirming. Also, thanks again for flagging this bug.

All current packages are made in Docker containers running either on a Fedora VM when I build them locally or on an Ubuntu VM for those built in Azure DevOps. Here they are, if you want to build it locally.

https://hub.docker.com/u/stonesteps

You can see how each is set up in their Dockerfile here:

https://github.com/StoneStepsInc/StoneStepsWebalizer/tree/master/devops

In Azure DevOps they are built via their pipeline, which you can see here (scroll to the section with the comment `Linux`):

https://github.com/StoneStepsInc/StoneStepsWebalizer/blob/master/azure-pipelines.yml

The build against the main branch is automatic, and in case if you want to try out some a CI build, you can find binaries for all platforms in Azure Pipelines (section Related):

https://dev.azure.com/StoneSteps/Stone%20Steps%20Webalizer/_build?definitionId=6

As for your changes in `sample.conf`, they are in `master` and the release is a patch release for the v6.0 line, which is on the `v6-0` branch, which does not contain your changes. When v6.1.0 is released, which is currently `master`, these changes will be in that release.

**Comment by IzzySoft** [comment]
> As for your changes in `sample.conf`

OK, got it – thanks! Then I'll release it as-is now, and leave the conf updates for v6.1.0.

> if you want to build it locally.

I can't. My question was rather: up to now I always used the Debian build for packaging. As there was no separate Ubuntu build I assumed it would work on other Debian derivatives "just the same" (I'm only running it on Debian myself). Now for the first time there's a separate Ubuntu build. Does that mean the Debian one no longer works on Ubuntu – or just that the Ubuntu one is "better optimized" in some way for Ubuntu?

I'll just package the Debian one now, as usual (as that's also the one I use) – and wait what comes :wink:

**Comment by StoneStepsInc** [comment]
The difference between Linux builds will be in versions of build tools and dependencies, such as the GCC C++ compiler:

* Ubuntu GCC C++ v9.3
* Debian GCC C++ v8.3

Other packages, such as Berkeley DB, MaxMind and GD, are also different, so it depends how their libraries behave when build tools and dependencies at build time and run time may be different.

I would say (and it's a speculation on my part) that if a build made with older tools on a newer OS runs at all, it likely means that it will run fine, as people usually try to keep older interfaces more or less stable. So, it sounds like a Debian build should run on Ubuntu, although I never tested these cross-flavor runs.

**Comment by IzzySoft** [comment]
OK, I'd say we wait if there are any complaints – and decide what action to take should it become necessary. Nobody complained in the past, so I don't see why that should start now. I've already pushed v6.0.1 to my repo, so it's available for those interested. As for "a build made with older tools on a newer OS": I have some binaries I compiled ages ago which are still running fine on rather recent systems, so I guess there won't be any complaints at all. And should there be such a case, a "quick fix" should be as easy as replacing the executable from the other tarball, right?

A big thanks to you for all the efforts taken! Processing seems to work well again on my server, so I guess we can close this issue.
